### PR TITLE
[watermarking] Add SBW (Stateless Bernoulli Watermarking) algorithm

### PR DESCRIPTION
**Pull request by si-mon-jinn**
## Purpose

Adds SBW (Stateless Bernoulli Watermarking) as a second watermarking algorithm
alongside the existing Gumbel-max scheme introduced in #54053 and extended in #56122.

Reference: Ceppi & Sanchez, "Flip, Don't Shuffle: Watermarking LLMs at the
Speed of Inference", EMNLP 2026. https://arxiv.org/abs/2609.03844

SBW is a KGW-family (Kirchenbauer et al.) logit-bias watermark. Like KGW, it
derives a pseudorandom green list from the token context and adds a bias `+delta`
to green token logits before sampling. Unlike KGW, green-list membership is
determined per token independently via a single counter-based PRF call
(Philox 4x32-10), with no vocabulary partitioning or hash list. This makes
the scheme stateless and O(1) per (token, context) pair. The detector
re-derives the same green lists from context and runs a z-score test over the
fraction of green tokens in the output.

The Philox 4x32-10 PRF is implemented from scratch over int32 to match the
paper's integer-domain greenness test exactly (`CBRNG(v, seed_t) < gamma *
INT31_MAX`). The same implementation is used in both the sampler and the detector,
ensuring bit-identical results across devices and releases. Detection requires
only the output token IDs and the watermark parameters; no model weights are needed.

Two seeding schemes are supported via `sbw_scheme`:
- `selfhash` (default): anchored minhash PRF, seed depends on context and the
  candidate token. More robust to substitution attacks. Default `context_width=4`.
- `lefthash`: additive PRF, one seed per context row. Faster. Default `context_width=1`.

Key parameters: `sbw_gamma` (green-list fraction, default 0.5), `sbw_delta`
(logit bias for green tokens, default 2.0).

SBW is compatible with greedy decoding and has native speculative-decoding support.

This PR also adds a `supports_greedy` property to the `Watermarker` ABC. SBW
biases logits rather than adding noise, so greedy requests (`temperature=0`)
can be watermarked normally. `GPUWatermarkSampler` branches on this property;
Gumbel and dual-key Gumbel keep `False` (behavior unchanged).

Note on implementation architecture: `GPUWatermarkSampler` currently passes
pre-top-k/top-p logits to `watermarker.sample()`, which fits noise-based
watermarkers like Gumbel naturally. SBW requires the bias to be applied on raw
logits (before temperature scaling and top-k/top-p filtering) to match the
paper's semantics and the standalone [vllm-sbw](https://github.com/si-mon-jinn/sbw)
reference implementation. This PR handles this via a `sample()` override in
`GPUWatermarkSampler` that intercepts raw logits for bias-based watermarkers
(`supports_greedy=True`). A cleaner long-term architecture would have the
`Watermarker` ABC own the full pipeline from raw logits to token ids, making
`GPUWatermarkSampler` fully watermarker-agnostic. Happy to implement this
generalization here or as a follow-up, based on reviewer preference.

## Test Plan

```bash
pytest tests/watermarking/test_sbw.py tests/watermarking/test_watermarking.py -v
```

- `tests/watermarking/test_sbw.py` (new, 40 tests): config/validation, PRF
  contracts (determinism, context/key/position sensitivity), CUDA parity
  (CPU and GPU produce bit-identical results), detection (watermarked text
  detected at p<0.05, wrong-key rejection), GPU sampler bias path (raw logit
  interception, mixed-batch per-row delta, greedy request handling).
- `tests/watermarking/test_watermarking.py`: updated for `supports_greedy`
  branching and mixed-batch behavior.

## Test Result

Measured on RTX 3090, Qwen3-8B (non-thinking mode), temp=0.6, top_k=20, top_p=0.95
(Qwen3 recommended defaults), 3x512 prompts, 30 input / 200 output tokens.

### Throughput

| Mode           | conc=32 (tok/s) | overhead | conc=64 (tok/s) | overhead | conc=128 (tok/s) | overhead |
|----------------|-----------------|----------|-----------------|----------|------------------|----------|
| base           | 1088 +/- 33     |          | 1526 +/- 3      |          | 1418 +/- 1       |          |
| SBW            | 1075 +/- 25     | +1.2%    | 1520 +/- 2      | +0.4%    | 1414 +/- 4       | +0.3%    |
| Gumbel         | 1080 +/- 29     | +0.8%    | 1531 +/- 1      | -0.3%    | 1412 +/- 63      | +0.4%    |
| DualKey Gumbel | 1079 +/- 24     | +1.0%    | 1523 +/- 1      | +0.1%    | 1419 +/- 2       | -0.1%    |

All three algorithms are within measurement noise of baseline at all concurrencies.

### Detection quality and text quality (100 C4 samples)

Perplexity scored by Qwen3.5-27B-FP8. Diversity is the n-gram diversity score
(higher = less repetitive). Detection at p_value_threshold=0.01 with context
deduplication enabled.

| Algorithm      | TPR  | FPR  | z_wm | PPL wm (median) | PPL unwm (median) | Diversity wm | Diversity unwm |
|----------------|------|------|------|-----------------|-------------------|--------------|----------------|
| SBW            | 0.96 | 0.04 | 5.09 | 4.06            | 3.48              | 0.482        | 0.525          |
| Gumbel         | 0.84 | 0.03 | 4.02 | 3.23            | 3.40              | 0.387        | 0.513          |
| DualKey Gumbel | 0.83 | 0.00 | 3.80 | 3.44            | 3.40              | 0.503        | 0.513          |

SBW achieves the highest TPR (0.96) and z-score (5.09) at a modest perplexity
cost (~17% median) from the logit bias. Single-key Gumbel produces lower
perplexity than baseline but reduces diversity by ~25%. DualKey Gumbel
(introduced in #56122) recovers the diversity loss and is perplexity-neutral
with zero FPR, at the cost of not supporting greedy decoding.

---

Builds on #54053 (Gumbel watermarking) and #56122 (dual-key Gumbel + context
deduplication). Relates to the watermarking hardening tracking issue #56105.

cc @TQCB as author of #54053 and #56122

*Parts of this implementation were developed with AI assistance (Claude Sonnet 4.6). All code has been reviewed and validated by the author.*

### CODE REVIEW

**Comment by si-mon-jinn** [comment]
Hi, friendly ping on this PR. It adds SBW as a second watermarking algorithm alongside Gumbel.

@LucasWilkinson @NickLucche: tagging you both since you reviewed the prior watermarking PRs (#54053, #56122). Not sure if either of you is the right person for this one, but you have the most context on this area. Happy to be redirected if there's someone else who should look at it.
  
The PR includes 40 new tests (all passing locally) and benchmarks showing no measurable throughput overhead. Let me know if there's anything I should address first.

Thanks!
